### PR TITLE
Create a new wrapper class for worn items

### PR DESCRIPTION
**Pull request by KorGgenT**
#### Summary
Infrastructure "Create a new wrapper class for worn items"

#### Purpose of change
A wrapper class is necessary, because you can change how something interacts with the concept of clothing, without finding all of the locations where the code goes `for( const item &clothing : you.worn )` and changing each instance of that. This is also step 1 of fixing up the clothing mechanics for the new limb mechanics. Once the wrapper class is in place, it should be a little easier to actually review the code changes necessary to make clothing work properly for arbitrary limbs.

#### Describe the solution
1) create new class `outfit`
2) change Character::worn to type `outfit`
3) address all red squiggly lines

#### Describe alternatives you've considered
I suppose it could be ad-hoc? but that's not a _good_ alternative.

#### Testing
in additon to unit tests not showing regressions, the save/load cycle over versions and within this version should be tested thoroughly. 

The most important test is save/load cycles keeping the clothing.

#### Additional context
This brings back the memory of how difficult it was to do inventory, but on a smaller scale. I'm opening up a draft PR mostly so i can keep an eye on inevitable conflicts. I've already had to nuke this branch once due to some systemic change with `worn` items. I also wanted to put up this draft early so that people will know what i'm working on.

special thanks to BombasticSlacks for helping me out!

### CODE REVIEW

**Comment by bombasticSlacks** [comment]
```

diff --git a/src/armor_layers.cpp b/src/armor_layers.cpp
index 612ad7581f5..2d179f9e4f9 100644
--- a/src/armor_layers.cpp
+++ b/src/armor_layers.cpp
@@ -320,8 +320,6 @@ std::vector<std::string> clothing_flags_description( const item &worn_item )
 item_penalties outfit::get_item_penalties( std::list<item>::const_iterator worn_item_it,
         const Character &c, const bodypart_id &_bp )
 {
-    layer_level layer = worn_item_it->get_layer();
-
     std::vector<bodypart_id> body_parts_with_stacking_penalty;
     std::vector<bodypart_id> body_parts_with_out_of_order_penalty;
     std::vector<std::set<std::string>> lists_of_bad_items_within;
@@ -333,20 +331,47 @@ item_penalties outfit::get_item_penalties( std::list<item>::const_iterator worn_
         if( !worn_item_it->covers( bp ) ) {
             continue;
         }
-        const int num_items = std::count_if( worn.begin(), worn.end(),
-        [layer, bp]( const item & i ) {
-            return i.get_layer() == layer && i.covers( bp ) && !i.has_flag( flag_SEMITANGIBLE );
-        } );
-        if( num_items > 1 ) {
-            body_parts_with_stacking_penalty.push_back( bp );
-        }
-
         std::set<std::string> bad_items_within;
-        for( auto it = worn.begin(); it != worn_item_it; ++it ) {
-            if( it->get_layer() > layer && it->covers( bp ) ) {
-                bad_items_within.insert( it->type_name() );
+        // if no subparts do the old way
+        if( bp->sub_parts.empty() ) {
+            std::vector<layer_level> layer = worn_item_it->get_layer( bp );
+            const int num_items = std::count_if( worn.begin(), worn.end(),
+            [layer, bp]( const item & i ) {
+                return i.has_layer( layer, bp ) && i.covers( bp ) && !i.has_flag( flag_SEMITANGIBLE );
+            } );
+            if( num_items > 1 ) {
+                body_parts_with_stacking_penalty.push_back( bp );
+            }
+            for( auto it = worn.begin(); it != worn_item_it; ++it ) {
+                if( it->get_layer( bp ) > layer && it->covers( bp ) ) {
+                    bad_items_within.insert( it->type_name() );
+                }
             }
+        } else {
+            for( const auto &sbp : bp->sub_parts ) {
+                if( !worn_item_it->covers( sbp ) ) {
+                    continue;
+                }
+                std::vector<layer_level> layer = worn_item_it->get_layer( sbp );
+                const int num_items = std::count_if( worn.begin(), worn.end(),
+                [layer, bp, sbp]( const item & i ) {
+                    return i.has_layer( layer, sbp ) && i.covers( bp ) && !i.has_flag( flag_SEMITANGIBLE ) &&
+                           i.covers( sbp );
+                } );
+                if( num_items > 1 ) {
+                    body_parts_with_stacking_penalty.push_back( bp );
+                }
+                for( auto it = worn.begin(); it != worn_item_it; ++it ) {
+                    if( it->get_layer( sbp ) > layer && it->covers( sbp ) ) {
+                        bad_items_within.insert( it->type_name() );
+                    }
+                }
+            }
+
         }
+
+
+
         if( !bad_items_within.empty() ) {
             body_parts_with_out_of_order_penalty.push_back( bp );
             lists_of_bad_items_within.push_back( bad_items_within );
diff --git a/src/character_attire.cpp b/src/character_attire.cpp
index 8531d9299da..3c0b4916395 100644
--- a/src/character_attire.cpp
+++ b/src/character_attire.cpp
@@ -823,12 +823,13 @@ bool outfit::worn_with_flag( const flag_id &f ) const
 bool outfit::is_worn_item_visible( std::list<item>::const_iterator worn_item,
                                    const body_part_set &worn_item_body_parts ) const
 {
+    const body_part_set worn_item_body_parts = worn_item->get_covered_body_parts();
     return std::any_of( worn_item_body_parts.begin(), worn_item_body_parts.end(),
     [this, &worn_item]( const bodypart_str_id & bp ) {
         // no need to check items that are worn under worn_item in the armor sort order
         for( auto i = std::next( worn_item ), end = worn.end(); i != end; ++i ) {
-            if( i->covers( bp ) && i->get_layer() != layer_level::BELTED &&
-                i->get_layer() != layer_level::WAIST &&
+            if( i->covers( bp ) &&
+                !i->has_layer( { layer_level::BELTED, layer_level::WAIST } ) &&
                 i->get_coverage( bp ) >= worn_item->get_coverage( bp ) ) {
                 return false;
             }
```

**Comment by bombasticSlacks** [comment]
```diff
diff --git a/src/character_attire.cpp b/src/character_attire.cpp
index a47aaca6729..58ae24de790 100644
--- a/src/character_attire.cpp
+++ b/src/character_attire.cpp
@@ -1544,6 +1544,20 @@ static void item_armor_enchantment_adjust( Character &guy, damage_unit &du, item
 void outfit::absorb_damage( Character &guy, damage_unit &elem, bodypart_id bp,
                             std::list<item> &worn_remains, bool &armor_destroyed )
 {
+    sub_bodypart_id sbp;
+    sub_bodypart_id secondary_sbp;
+    // if this body part has sub part locations roll one
+    if( !bp->sub_parts.empty() ) {
+        sbp = bp->random_sub_part( false );
+        // the torso has a second layer of hanging body parts
+        if( bp == body_part_torso ) {
+            secondary_sbp = bp->random_sub_part( true );
+        }
+    }
+
+    // generate a single roll for determining if hit
+    int roll = rng( 1, 100 );
+
     // Only the outermost armor can be set on fire
     bool outermost = true;
     // The worn vector has the innermost item first, so
@@ -1574,7 +1588,18 @@ void outfit::absorb_damage( Character &guy, damage_unit &elem, bodypart_id bp,
         }
 
         if( !destroy ) {
-            destroy = guy.armor_absorb( elem, armor, bp );
+            // if we don't have sub parts data
+            // this is the feet head and hands
+            if( bp->sub_parts.empty() ) {
+                destroy = guy.armor_absorb( elem, armor, bp, roll );
+            } else {
+                // if this armor has sublocation data test against it instead of just a generic roll
+                destroy = guy.armor_absorb( elem, armor, bp, sbp, roll );
+                // for the torso we also need to consider if it hits anything hanging off the character or their neck
+                if( bp == body_part_torso ) {
+                    destroy = guy.armor_absorb( elem, armor, bp, secondary_sbp, roll );
+                }
+            }
         }
 
         if( destroy ) {
```

**Comment by bombasticSlacks** [comment]
are you waiting on anything more from me to get this working?

**Comment by KorGgenT** [comment]
OK! It's ready! it runs! it saves! it loads! it saves and loads again!

File: src/character_attire.cpp [hunk 1]
```diff
@@ -892,3 +770,1270 @@ void Character::item_encumb( std::map<bodypart_id, encumbrance_data> &vals,
...
+        std::advance( iter, position );
+        return *iter;
+    } else {
+        return null_item_reference();
+    }
+}
+
+std::string outfit::get_armor_display( bodypart_id bp, unsigned int truncate ) const
+{
+    for( const auto it = worn.rbegin(); it != worn.rend(); ) {
```
Comment (akrieger) [inline src/character_attire.cpp:1963]: Infinite loop.
```suggestion
    for( auto it = worn.rbegin(); it != worn.rend(); ++it ) {
```

**Pull request merged**